## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [2]:
import pandas as pd

df = pd.read_csv('client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [3]:
# View the column names in the data
df.columns


Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [4]:
# Use the describe function to gather some basic statistics
df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [4]:
# Use this space to do any additional research
# and familiarize yourself with the data.


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   first        54639 non-null  object 
 1   last         54639 non-null  object 
 2   job          54639 non-null  object 
 3   phone        54639 non-null  object 
 4   email        54639 non-null  object 
 5   client_id    54639 non-null  int64  
 6   order_id     54639 non-null  int64  
 7   order_date   54639 non-null  object 
 8   order_week   54639 non-null  int64  
 9   order_year   54639 non-null  int64  
 10  item_id      54639 non-null  object 
 11  category     54639 non-null  object 
 12  subcategory  54639 non-null  object 
 13  unit_price   54639 non-null  float64
 14  unit_cost    54639 non-null  float64
 15  unit_weight  54639 non-null  float64
 16  qty          54639 non-null  int64  
 17  line_number  54639 non-null  int64  
dtypes: float64(3), int64(6), object(9)
memory usag

In [6]:
# What three item categories had the most entries?
df['category'].value_counts().head(3)


category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [8]:
# For the category with the most entries, which subcategory had the most entries?
df['subcategory'].value_counts().head(1)

subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [9]:
# Which five clients had the most entries in the data?
df['client_id'].value_counts().head(5)

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [10]:
# Store the client ids of those top 5 clients in a list.
top_5 = df['client_id'].value_counts().head(5).index.tolist()
top_5

[33615, 66037, 46820, 38378, 24741]

In [11]:
# How many total units (the qty column) did the client with the most entries order order?
df.loc[df['client_id'] == top_5[0]]['qty'].sum()

64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [12]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
df['subtotal'] = df['unit_price'] * df['qty']
df[['unit_price','qty','subtotal']].head()

unit_price  qty   subtotal
0     1096.80  105  115164.00
1       24.95   21     523.95
2       13.52   39     527.28
3       36.42   29    1056.18
4      195.10   20    3902.00

In [14]:
# Create a column for shipping price. Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
def final_shipping_price(x):
    if x > 50:
        return 7
    else:
        return 10

df['total_weight'] = df['qty'] * df['unit_weight']
df['shipping_price'] = df['total_weight'] * df['total_weight'].apply(final_shipping_price)
df[['unit_price','unit_weight','qty','total_weight','shipping_price']].head()




unit_price  unit_weight  qty  total_weight  shipping_price
0     1096.80         7.50  105        787.50         5512.50
1       24.95         1.49   21         31.29          312.90
2       13.52         1.68   39         65.52          458.64
3       36.42         1.23   29         35.67          356.70
4      195.10        46.43   20        928.60         6500.20

In [15]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
df['total_price'] = df['subtotal'] + df['shipping_price'] + (df['subtotal'] + df['shipping_price']) * 0.0925
df[['subtotal','shipping_price','total_price']].head()

subtotal  shipping_price    total_price
0  115164.00         5512.50  131839.076250
1     523.95          312.90     914.258625
2     527.28          458.64    1077.117600
3    1056.18          356.70    1543.571400
4    3902.00         6500.20   11364.403500

In [16]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
df['line_cost'] = df['unit_cost'] * df['qty'] + df['shipping_price']
df[['unit_cost','qty','shipping_price','line_cost']].head()


unit_cost  qty  shipping_price  line_cost
0     762.71  105         5512.50   85597.05
1      15.09   21          312.90     629.79
2       7.86   39          458.64     765.18
3      24.85   29          356.70    1077.35
4     108.17   20         6500.20    8663.60

In [17]:
# Create a column for the profit of each line using line cost and line price
df['line_profit'] = df['total_price'] - df['line_cost']
df[['line_cost','total_price','line_profit']].head()



line_cost    total_price   line_profit
0   85597.05  131839.076250  46242.026250
1     629.79     914.258625    284.468625
2     765.18    1077.117600    311.937600
3    1077.35    1543.571400    466.221400
4    8663.60   11364.403500   2700.803500

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [18]:
# Check your work using the totals above
order_check = [2742071,2173913,6128929]

order_group_df = df.groupby('order_id').agg({'total_price':'sum'}).reset_index()

order_check_df = order_group_df[order_group_df['order_id'].isin(order_check)]

for order in order_check:
    print(f"Order {order} total: ${order_check_df.loc[order_check_df['order_id'] == order]['total_price'].values[0]:,.2f}")

Order 2742071 total: $152,811.90
Order 2173913 total: $162,388.72
Order 6128929 total: $923,441.24


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [19]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
top_5_df = df.groupby('client_id').agg({'total_price':'sum'}).reset_index()
top_5_df[top_5_df['client_id'].isin(top_5)]



client_id   total_price
144      24741  8.226889e+07
258      33615  8.377309e+06
319      38378  1.290655e+07
402      46820  9.743794e+06
621      66037  1.025951e+07

In [20]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 

top_5_summary = df.groupby('client_id').agg({'qty':'sum','shipping_price':'sum','total_price':'sum','line_profit':'sum'}).reset_index()
top_5_summary[top_5_summary['client_id'].isin(top_5)]


client_id     qty  shipping_price   total_price   line_profit
144      24741  239862      5126448.37  8.226889e+07  3.657999e+07
258      33615   64313      1828984.89  8.377309e+06  2.201995e+06
319      38378   73667      3429455.40  1.290655e+07  3.271830e+06
402      46820   75768      1601448.84  9.743794e+06  2.736603e+06
621      66037   43018      1395151.85  1.025951e+07  3.255032e+06

In [28]:
# Format the data and rename the columns to names suitable for presentation.
top_5_summary.columns = ['Client ID','Total Units Purchased','Total Shipping Price','Total Revenue','Total Profit'] 
top_5_summary['Total Units Purchased'] = top_5_summary['Total Units Purchased'].apply(lambda x: "{x:,.0f}")
top_5_summary['Total Shipping Price'] = top_5_summary['Total Shipping Price'].apply(lambda x: "${x:,.2f}")
top_5_summary['Total Revenue'] = top_5_summary['Total Revenue'].apply(lambda x: "${x:,.2f}")
top_5_summary['Total Profit'] = top_5_summary['Total Profit'].apply(lambda x: "${x:,.2f}")
top_5_summary

# Define the money columns. 
money_columns = ['unit_price','unit_cost','subtotal','shipping_price','total_price','line_cost','line_profit']

# Define a function that converts a dollar amount to millions.
def currency_format_millions(x):
    return "${x/1000000:,.2f}M"

# Apply the currency_format_millions function to only the money columns. 
df[money_columns] = df[money_columns].applymap(currency_format_millions)

# Rename the columns to reflect the change in the money format. 
df.columns = ['Order ID', 'Client ID', 'Category', 'Subcategory', 'Product', 'Unit Price', 'Unit Cost', 'Unit Weight', 'Qty', 'Subtotal', 'Total Weight', 'Shipping Price', 'Total Price', 'Line Cost', 'Line Profit']



/tmp/ipykernel_5643/2130444533.py:17: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  df[money_columns] = df[money_columns].applymap(currency_format_millions)


ValueError: Length mismatch: Expected axis has 24 elements, new values have 15 elements

In [29]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
df.sort_values('Line Profit', ascending=False)


KeyError: 'Line Profit'